### PR TITLE
Species tree pipeline: cache annotations

### PR DESCRIPTION
**Pull request by ens-sb**
## Description

Added functionality to cache annotations in the results directory.

**Related JIRA tickets:**
- ENSCOMPARASW-5726

## Testing
The pipeline was run on a  set of rice species.
---

## PR review checklist

- [ ] Is the PR against an appropriate branch?
- [ ] Does the code adhere to coding guidelines?
- [ ] Does the code do what it claims to do?
- [ ] Is the code readable? Is it appropriately documented?
- [ ] Is the logic in the correct place?
- [ ] Was the code tested appropriately? Are there unit tests? Are unit tests self-contained and non-redundant?
- [ ] Did Travis CI pass for the code in the PR? Is Codecov acceptable based on the included/updated unit tests?
- [ ] Will the new code fail gracefully?
- [ ] Does the code follow good practice for writing performant code (e.g. using a database transaction rather than repeated queries outside of a transaction)?
- [ ] Does it bring in an unnecessary dependency?
- [ ] If you are reviewing a new analysis, is it future-proof and pluggable?
- [ ] Does the PR meet agile guidelines?

### CODE REVIEW

File: pipelines/SpeciesTreeFromBusco/main.nf [hunk 1]
```diff
@@ -86,6 +86,30 @@ process prepareBusco {
     """
 }
 
+
+/**
```
Comment (twalsh-ebi) [inline pipelines/SpeciesTreeFromBusco/main.nf:90]: The workflow now includes a lot of comments describing what the processes do. Where those comments are a general description of the process, I'd suggest to move them to the Groovydoc of the process itself. Here is an illustration of what I have in mind:
```suggestion
/**
* Link annotations present in the cache.
*
```

File: pipelines/SpeciesTreeFromBusco/main.nf [hunk 2]
```diff
@@ -292,18 +322,56 @@ process runIqtree {
...
-    buscoAnnot(prepareBusco.out.busco_prots, prepareGenome.out.proc_genome)
-    runGffread(buscoAnnot.out.busco_annot, buscoAnnot.out.genome)
+    // Branch the genomes channel based on presence in the
+    // annotation cache:
+    prepareGenome.out.proc_genome.branch {
+        annot: !annoInCache(it, params.anno_cache)
+        link: annoInCache(it, params.anno_cache)
+    }.set { genome_fork }
+
+    // Link annotations present in the cache:
```
Comment (twalsh-ebi) [inline pipelines/SpeciesTreeFromBusco/main.nf:351]: If you decide to put this description in the Groovydoc, I'd suggest to remove it from here.
```suggestion
```

File: pipelines/SpeciesTreeFromBusco/main.nf [hunk 3]
```diff
@@ -86,6 +86,30 @@ process prepareBusco {
     """
 }
 
+
+/**
+*@input path to the genome fasta
+*@output path to annotation GTF
+*@output path to genome fasta
+*/
+process linkAnnoCache {
```
Comment (twalsh-ebi) [inline pipelines/SpeciesTreeFromBusco/main.nf:95]: If `params.anno_cache` is set to `${params.results_dir}/anno_cache`, what does the `linkAnnoCache` process do?

File: pipelines/SpeciesTreeFromBusco/main.nf [hunk 4]
```diff
@@ -292,18 +322,56 @@ process runIqtree {
     """
 }
 
+// Function to check if a genome is present in
+// the annotation cache.
+def annoInCache(genome, anno_cache) {
```
Comment (twalsh-ebi) [inline pipelines/SpeciesTreeFromBusco/main.nf:327]: Does the annotation cache need to handle updated input genomes? For example, if for some reason a genome FASTA file is updated, is it forwarded to `linkAnnoCache` or `buscoAnnot`?

File: pipelines/SpeciesTreeFromBusco/main.nf [hunk 5]
```diff
@@ -292,18 +322,56 @@ process runIqtree {
...
+    }.set { genome_fork }
+
+    // Link annotations present in the cache:
+    linkAnnoCache(genome_fork.link)
+    // Annotate genomes not present in the cache:
+    buscoAnnot(prepareBusco.out.busco_prots, genome_fork.annot)
+
+    // Merge the output of link and annotate steps:
+    genomes = buscoAnnot.out.genome.mix(linkAnnoCache.out.genome)
+    annots = buscoAnnot.out.busco_annot.mix(linkAnnoCache.out.busco_annot)
```
Comment (twalsh-ebi) [inline pipelines/SpeciesTreeFromBusco/main.nf:358]: Do `genomes` and `annots` come out of the mix OK?

File: pipelines/SpeciesTreeFromBusco/main.nf [hunk 6]
```diff
@@ -51,8 +51,8 @@ process prepareGenome {
     output:
         path "processed/*", emit: proc_genome
     script:
-        id = (genome =~ /(.*)\..*$/)[0][1]
-        ext = (genome =~ /.*\.(.*)$/)[0][1]
+        id = (genome =~ /(.*)\..*?$/)[0][1]
```
[the remark this answers is not in the document]
Comment (ens-sb) [reply pipelines/SpeciesTreeFromBusco/main.nf:54]: Yes, the plus sign is probably better here.

File: pipelines/SpeciesTreeFromBusco/main.nf [hunk 7]
```diff
@@ -51,8 +51,8 @@ process prepareGenome {
...
+        ext = (genome =~ /.*\.(.*?)$/)[0][1]
```
[the remark this answers is not in the document]
Comment (ens-sb) [reply pipelines/SpeciesTreeFromBusco/main.nf:55]: Yes, it is sensible to remove the ext variable.

File: pipelines/SpeciesTreeFromBusco/main.nf (hunk 3 above)
Comment (ens-sb) [reply pipelines/SpeciesTreeFromBusco/main.nf:95]: The linkAnnoCache process reuses genome annotations from the specified cache directory.

File: pipelines/SpeciesTreeFromBusco/main.nf (hunk 4 above)
Comment (ens-sb) [reply pipelines/SpeciesTreeFromBusco/main.nf:327]: No, the name of the fasta has to be changed after update or deleted from the cache to trigger reannotation.

File: pipelines/SpeciesTreeFromBusco/main.nf (hunk 5 above)
Comment (ens-sb) [reply pipelines/SpeciesTreeFromBusco/main.nf:358]: Yes, this is a serious issue. Will have to pass around tuples to solve the potential mismatch in order.

File: pipelines/SpeciesTreeFromBusco/main.nf (hunk 1 above)
Comment (ens-sb) [reply pipelines/SpeciesTreeFromBusco/main.nf:90]: I would like to keep those comments close to the call of the process. I will consider adding a groovydoc for the workflow as the pipeline matures.

File: pipelines/SpeciesTreeFromBusco/main.nf (hunk 4 above)
Comment (twalsh-ebi) [reply pipelines/SpeciesTreeFromBusco/main.nf:327]: OK, thanks. I was trying to get a sense of what the cache is expected to do.

File: pipelines/SpeciesTreeFromBusco/main.nf (hunk 5 above)
Comment (twalsh-ebi) [reply pipelines/SpeciesTreeFromBusco/main.nf:358]: I understand this is to be addressed in an imminent pull request.

File: pipelines/SpeciesTreeFromBusco/main.nf (hunk 3 above)
Comment (twalsh-ebi) [reply pipelines/SpeciesTreeFromBusco/main.nf:94]: I'm thinking about this specific case, but from our discussion outside the PR I understand that this is not how the anno_cache is expected to be used.

**Review by twalsh-ebi — APPROVED** [review]

**Pull request merged**